File: packages/sdk-node/src/global-accounts.ts
// ============================================================
// @kangopenbanking/sdk — GlobalAccountsResource (Nium-powered)
// Aligned with OpenAPI v4.50.0 — /v1/gateway/global-accounts*
// ============================================================

import type { KangOpenBanking } from './client';

export type GlobalAccountCurrency = 'USD' | 'EUR' | 'GBP';
export type GlobalAccountStatus = 'active' | 'suspended' | 'closed';
export type PayoutPreference = 'KANG_WALLET' | 'MOBILE_MONEY';
export type NiumMode = 'stub' | 'sandbox' | 'live';

export interface NiumGlobalAccount {
  id: string;
  currency: GlobalAccountCurrency;
  iban: string | null;
  account_number: string | null;
  routing_code: string | null;
  bic: string | null;
  bank_name: string;
  bank_address: string | null;
  beneficiary_name: string;
  status: GlobalAccountStatus;
  payout_preference_override: PayoutPreference | null;
  payout_channel_override: string | null;
  mode: NiumMode;
}

export interface NiumIncomingPayment {
  id: string;
  source_amount: number;
  source_currency: string;
  fx_rate_nium: number;
  fx_spread_bps: number;
  xaf_gross: number;
  xaf_spread_revenue: number;
  xaf_withdrawal_fee: number;
  xaf_net_credited: number;
  routing: PayoutPreference;
  status:
    | 'received'
    | 'credited'
    | 'payout_pending'
    | 'payout_completed'
    | 'payout_failed'
    | 'failed';
  created_at: string;
}

export interface NiumPayoutPreferenceDefaults {
  payout_preference: PayoutPreference;
  payout_channel: string | null;
}

export interface ListGlobalAccountsResponse {
  accounts: NiumGlobalAccount[];
  incoming_payments: NiumIncomingPayment[];
  user_defaults: NiumPayoutPreferenceDefaults;
}

export interface CreateGlobalAccountRequest {
  currency: GlobalAccountCurrency;
  beneficiary_name?: string;
}

export interface CreateGlobalAccountResponse {
  account: NiumGlobalAccount;
  reused: boolean;
}

export type UpdatePayoutPreferenceRequest =
  | {
      scope: 'user';
      payout_preference: PayoutPreference;
      /** Required when payout_preference = MOBILE_MONEY (E.164 phone). */
      payout_channel?: string;
    }
  | {
      scope: 'account';
      account_id: string;
      /** Pass null to clear the per-account override. */
      payout_preference_override?: PayoutPreference | null;
      payout_channel_override?: string | null;
    };

export class GlobalAccountsResource {
  constructor(private client: KangOpenBanking) {}

  /** POST /v1/gateway/global-accounts — idempotent per (user, currency). */
  async create(params: CreateGlobalAccountRequest): Promise<CreateGlobalAccountResponse> {
    return this.client.request('POST', 'nium-create-global-account', params);
  }

  /** GET /v1/gateway/global-accounts — accounts + recent incoming payments + user defaults. */
  async list(): Promise<ListGlobalAccountsResponse> {
    return this.client.request('GET', 'nium-list-global-accounts');
  }

  /** PATCH /v1/gateway/global-accounts/payout-preference — user default or per-account override. */
  async updatePayoutPreference(params: UpdatePayoutPreferenceRequest): Promise<{ ok: true }> {
    return this.client.request('PATCH', 'nium-update-payout-preference', params);
  }

  /**
   * Verify a Nium webhook signature (HMAC-SHA256 of the raw request body,
   * sent in the `x-nium-signature` header).
   */
  static async verifyWebhookSignature(
    rawBody: string,
    signature: string,
    secret: string,
  ): Promise<boolean> {
    const enc = new TextEncoder();
    const key = await globalThis.crypto.subtle.importKey(
      'raw',
      enc.encode(secret),
      { name: 'HMAC', hash: 'SHA-256' },
      false,
      ['sign'],
    );
    const sig = await globalThis.crypto.subtle.sign('HMAC', key, enc.encode(rawBody));
    const hex = Array.from(new Uint8Array(sig))
      .map((b) => b.toString(16).padStart(2, '0'))
      .join('');
    if (hex.length !== signature.length) return false;
    let mismatch = 0;
    for (let i = 0; i < hex.length; i++) mismatch |= hex.charCodeAt(i) ^ signature.charCodeAt(i);
    return mismatch === 0;
  }
}

File: packages/sdk-python/kangopenbanking/global_accounts.py
"""GlobalAccountsResource — Nium-powered global virtual accounts.

Aligned with OpenAPI v4.50.0 — /v1/gateway/global-accounts*.
"""

from __future__ import annotations

import hashlib
import hmac
from dataclasses import dataclass
from typing import Any, Dict, List, Literal, Optional, Union


PayoutPreference = Literal["KANG_WALLET", "MOBILE_MONEY"]
GlobalAccountCurrency = Literal["USD", "EUR", "GBP"]


@dataclass
class NiumGlobalAccount:
    id: str
    currency: str
    bank_name: str
    beneficiary_name: str
    status: str
    mode: str
    iban: Optional[str] = None
    account_number: Optional[str] = None
    routing_code: Optional[str] = None
    bic: Optional[str] = None
    bank_address: Optional[str] = None
    payout_preference_override: Optional[str] = None
    payout_channel_override: Optional[str] = None


@dataclass
class NiumIncomingPayment:
    id: str
    source_amount: float
    source_currency: str
    fx_rate_nium: float
    fx_spread_bps: int
    xaf_gross: float
    xaf_spread_revenue: float
    xaf_withdrawal_fee: float
    xaf_net_credited: float
    routing: str
    status: str
    created_at: str


class GlobalAccountsResource:
    """Issue and manage Nium-powered USD / EUR / GBP receiving accounts."""

    def __init__(self, client):  # type: ignore[no-untyped-def]
        self._c = client

    def create(
        self,
        currency: GlobalAccountCurrency,
        beneficiary_name: Optional[str] = None,
    ) -> Dict[str, Any]:
        """POST /v1/gateway/global-accounts — idempotent per (user, currency)."""
        body: Dict[str, Any] = {"currency": currency}
        if beneficiary_name:
            body["beneficiary_name"] = beneficiary_name
        return self._c._request("POST", "nium-create-global-account", json=body)

    def list(self) -> Dict[str, Any]:
        """GET /v1/gateway/global-accounts — accounts, incoming payments, user defaults."""
        return self._c._request("GET", "nium-list-global-accounts")

    def update_payout_preference(
        self,
        scope: Literal["user", "account"],
        *,
        payout_preference: Optional[PayoutPreference] = None,
        payout_channel: Optional[str] = None,
        account_id: Optional[str] = None,
        payout_preference_override: Optional[PayoutPreference] = None,
        payout_channel_override: Optional[str] = None,
    ) -> Dict[str, Any]:
        """PATCH /v1/gateway/global-accounts/payout-preference."""
        body: Dict[str, Any] = {"scope": scope}
        if scope == "user":
            if payout_preference is None:
                raise ValueError("payout_preference is required when scope='user'")
            body["payout_preference"] = payout_preference
            if payout_channel is not None:
                body["payout_channel"] = payout_channel
        else:
            if not account_id:
                raise ValueError("account_id is required when scope='account'")
            body["account_id"] = account_id
            body["payout_preference_override"] = payout_preference_override
            body["payout_channel_override"] = payout_channel_override
        return self._c._request("PATCH", "nium-update-payout-preference", json=body)

    @staticmethod
    def verify_webhook_signature(payload: str, signature: str, secret: str) -> bool:
        """Verify the `x-nium-signature` header (HMAC-SHA256 of the raw body)."""
        computed = hmac.new(
            secret.encode("utf-8"), payload.encode("utf-8"), hashlib.sha256
        ).hexdigest()
        return hmac.compare_digest(computed, signature)
